File: src/core/gui/Dialog.ts
import { ResizableWidget } from "./ResizableWidget";
import { Widget, WidgetOptions } from "./Widget";
import { Child, div, h1, section } from "./dom";
import "./Dialog.css";
import { is_property, Property } from "../observable/property/Property";
import { WidgetProperty } from "../observable/property/WidgetProperty";
import { WritableProperty } from "../observable/property/WritableProperty";
import { Emitter } from "../observable/Emitter";
import { Observable } from "../observable/Observable";
import { emitter } from "../observable";

const DIALOG_WIDTH = 500;
const DIALOG_MAX_HEIGHT = 500;

export type DialogOptions = WidgetOptions & {
    readonly title?: string | Property<string>;
    readonly description?: string | Property<string>;
    readonly content?: Child | Property<Child>;
    readonly footer?: readonly Child[];
};

/**
 * A popup window with a title, description, body and dismiss button.
 */
export class Dialog extends ResizableWidget {
    private x = 0;
    private y = 0;
    private prev_mouse_x = 0;
    private prev_mouse_y = 0;

    private _title = new WidgetProperty<string>(this, "", this.set_title);
    private _description = new WidgetProperty<string>(this, "", this.set_description);
    private _content = new WidgetProperty<Child>(this, "", this.set_content);

    private readonly overlay_element: HTMLElement;
    private readonly header_element: HTMLElement;
    private readonly description_element: HTMLElement;
    private readonly content_element: HTMLElement;

    protected readonly _ondismiss: Emitter<Event> = emitter();

    readonly element: HTMLElement;
    readonly children: readonly Widget[] = [];

    readonly title: WritableProperty<string> = this._title;
    readonly description: WritableProperty<string> = this._description;
    readonly content: WritableProperty<Child> = this._content;

    /**
     * Emits an event when the user presses the escape key.
     */
    readonly ondismiss: Observable<Event> = this._ondismiss;

    constructor(options?: DialogOptions) {
        super(options);

        this.element = section(
            { className: "core_Dialog", tabIndex: 0 },
            (this.header_element = h1()),
            (this.description_element = div({ className: "core_Dialog_description" })),
            (this.content_element = div({ className: "core_Dialog_body" })),
            div({ className: "core_Dialog_footer" }, ...(options?.footer ?? [])),
        );

        this.element.style.width = `${DIALOG_WIDTH}px`;
        this.element.style.maxHeight = `${DIALOG_MAX_HEIGHT}px`;

        this.element.addEventListener("keydown", evt => this.keydown(evt));

        if (options) {
            if (typeof options.title === "string") {
                this.title.val = options.title;
            } else if (options.title) {
                this.title.bind_to(options.title);
            }

            if (typeof options.description === "string") {
                this.description.val = options.description;
            } else if (options.description) {
                this.description.bind_to(options.description);
            }

            if (is_property(options.content)) {
                this.content.bind_to(options.content);
            } else if (options.content != undefined) {
                this.content.val = options.content;
            }
        }

        this.set_position(
            (window.innerWidth - DIALOG_WIDTH) / 2,
            (window.innerHeight - DIALOG_MAX_HEIGHT) / 2,
        );

        this.header_element.addEventListener("mousedown", this.mousedown);

        this.overlay_element = div({ className: "core_Dialog_modal_overlay", tabIndex: -1 });
        this.overlay_element.addEventListener("focus", () => this.focus());

        this.finalize_construction();
    }

    dispose(): void {
        super.dispose();
        this.overlay_element.remove();
    }

    focus(): void {
        (this.first_focusable_child(this.element) || this.element).focus();
    }

    private first_focusable_child(element: HTMLElement): HTMLElement | undefined {
        for (const child of element.children) {
            if (child instanceof HTMLElement) {
                if (child.tabIndex >= 0) {
                    return child;
                } else {
                    const element = this.first_focusable_child(child);

                    if (element) {
                        return element;
                    }
                }
            }
        }

        return undefined;
    }

    set_position(x: number, y: number): void {
        this.x = x;
        this.y = y;
        this.element.style.transform = `translate(${Math.floor(x)}px, ${Math.floor(y)}px)`;
    }

    protected set_visible(visible: boolean): void {
        if (visible) {
            document.body.append(this.overlay_element);
            document.body.append(this.element);
            this.focus();
        } else {
            this.overlay_element.remove();
            this.element.remove();
        }
    }

    private set_title(title: string): void {
        this.header_element.textContent = title;
    }

    private set_description(description: string): void {
        if (description === "") {
            this.description_element.hidden = true;
            this.description_element.textContent = "";
        } else {
            this.description_element.hidden = false;
            this.description_element.textContent = description;
        }
    }

    private set_content(content: Child): void {
        this.content_element.textContent = "";
        this.content_element.append(content);
    }

    private mousedown = (evt: MouseEvent): void => {
        this.prev_mouse_x = evt.clientX;
        this.prev_mouse_y = evt.clientY;
        window.addEventListener("mousemove", this.window_mousemove);
        window.addEventListener("mouseup", this.window_mouseup);
    };

    private window_mousemove = (evt: MouseEvent): void => {
        evt.preventDefault();
        this.set_position(
            this.x + evt.clientX - this.prev_mouse_x,
            this.y + evt.clientY - this.prev_mouse_y,
        );
        this.prev_mouse_x = evt.clientX;
        this.prev_mouse_y = evt.clientY;
    };

    private window_mouseup = (evt: MouseEvent): void => {
        evt.preventDefault();
        window.removeEventListener("mousemove", this.window_mousemove);
        window.removeEventListener("mouseup", this.window_mouseup);
    };

    private keydown(evt: KeyboardEvent): void {
        if (evt.key === "Escape") {
            this._ondismiss.emit({ value: evt });
        }
    }
}

File: webui/src/jsMain/kotlin/world/phantasmal/webui/widgets/Dialog.kt
package world.phantasmal.webui.widgets

import kotlinx.browser.window
import org.w3c.dom.HTMLElement
import org.w3c.dom.Node
import org.w3c.dom.events.Event
import org.w3c.dom.events.KeyboardEvent
import org.w3c.dom.get
import org.w3c.dom.pointerevents.PointerEvent
import world.phantasmal.cell.Cell
import world.phantasmal.cell.emptyStringCell
import world.phantasmal.cell.isEmpty
import world.phantasmal.cell.trueCell
import world.phantasmal.webui.dom.div
import world.phantasmal.webui.dom.dom
import world.phantasmal.webui.dom.h1
import world.phantasmal.webui.dom.section

// TODO: Use HTML dialog element.
open class Dialog(
    visible: Cell<Boolean> = trueCell(),
    enabled: Cell<Boolean> = trueCell(),
    private val title: Cell<String>,
    private val description: Cell<String> = emptyStringCell(),
    private val content: Node.() -> Unit = {},
    private val footer: Node.() -> Unit = {},
    protected val onDismiss: () -> Unit = {},
) : Widget(visible, enabled) {
    private var x = 0
    private var y = 0

    private var overlayElement = dom {
        div {
            className = "pw-dialog-modal-overlay"
            tabIndex = -1

            addEventListener("focus", { this@Dialog.focus() })
        }
    }

    val dialogElement = dom {
        section {
            className = "pw-dialog"
            tabIndex = 0
            style.width = "${WIDTH}px"
            style.maxHeight = "${MAX_HEIGHT}px"

            addEventListener("keydown", ::onKeyDown)

            h1 {
                text(this@Dialog.title)

                onDrag(
                    onPointerDown = { true },
                    onPointerMove = ::onPointerMove,
                    onPointerUp = { it.preventDefault() },
                )
            }
            div {
                className = "pw-dialog-description"
                hidden(description.isEmpty())
                text(description)
            }
            div {
                className = "pw-dialog-body"
                content()
            }
            div {
                className = "pw-dialog-footer"
                footer()
            }
        }
    }

    init {
        observeNow(visible) {
            if (it) {
                setPosition(
                    (window.innerWidth - WIDTH) / 2,
                    (window.innerHeight - MAX_HEIGHT) / 2,
                )
                window.document.body!!.append(overlayElement)
                window.document.body!!.append(dialogElement)
                focus()
            } else {
                dialogElement.remove()
                overlayElement.remove()
            }
        }
    }

    override fun Node.createElement() = div { className = "pw-dialog-stub" }

    override fun dispose() {
        dialogElement.remove()
        overlayElement.remove()
        super.dispose()
    }

    private fun onPointerMove(movedX: Int, movedY: Int, e: PointerEvent): Boolean {
        e.preventDefault()
        setPosition(this.x + movedX, this.y + movedY)
        return true
    }

    private fun setPosition(x: Int, y: Int) {
        this.x = x
        this.y = y
        dialogElement.style.transform = "translate(${x}px, ${y}px)"
    }

    override fun focus() {
        (firstFocusableChild(dialogElement) ?: dialogElement).focus()
    }

    private fun firstFocusableChild(parent: HTMLElement): HTMLElement? {
        for (i in 0 until parent.children.length) {
            val child = parent.children[i]

            if (child is HTMLElement) {
                if (child.tabIndex >= 0) {
                    return child
                } else {
                    firstFocusableChild(child)?.let {
                        return it
                    }
                }
            }
        }

        return null
    }

    private fun onKeyDown(e: Event) {
        e as KeyboardEvent

        if (e.key == "Escape") {
            onDismiss()
        }
    }

    companion object {
        private const val WIDTH = 500
        private const val MAX_HEIGHT = 500

        init {
            @Suppress("CssUnresolvedCustomProperty", "CssUnusedSymbol")
            // language=css
            style("""
                .pw-dialog {
                    z-index: 20;
                    display: flex;
                    flex-direction: column;
                    outline: none;
                    position: fixed;
                    left: 0;
                    top: 0;
                    background-color: var(--pw-bg-color);
                    border: var(--pw-border);
                    padding: 10px;
                    box-shadow: black 0 0 10px -2px;
                }

                .pw-dialog:focus-within {
                    border: var(--pw-border-focus);
                }

                .pw-dialog h1 {
                    font-size: 20px;
                    margin: 0 0 10px 0;
                    padding-bottom: 4px;
                    border-bottom: var(--pw-border);
                }

                .pw-dialog-description {
                    user-select: text;
                    cursor: text;
                }

                .pw-dialog-body {
                    flex: 1;
                    margin: 4px 0;
                }

                .pw-dialog-footer {
                    display: flex;
                    flex-direction: row;
                    justify-content: flex-end;
                }

                .pw-dialog-footer > * {
                    margin-left: 2px;
                }

                .pw-dialog-modal-overlay {
                    outline: none;
                    z-index: 10;
                    position: fixed;
                    top: 0;
                    bottom: 0;
                    left: 0;
                    right: 0;
                    background-color: black;
                    opacity: 50%;
                    backdrop-filter: blur(5px);
                }
            """.trimIndent())
        }
    }
}
